x-pack/solutions/security/plugins/security_solution/server/lib/entity_analytics/privilege_monitoring/data_sources/sync/integrations/update_detection/queries.ts

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { estypes } from '@elastic/elasticsearch';
+import type { Matcher } from '../../../constants';
+import type { PrivilegeMonitoringDataClient } from '../../../../engine/data_client';
+import type { AfterKey } from './privileged_status_match';
+import type { PrivMonIntegrationsUser } from '../../../../types';
+import { makeIntegrationOpsBuilder } from '../../../bulk/upsert';
+import { errorsMsg, getErrorFromBulkResponse } from '../../utils';
+
+/**
+ * Build painless script for matcher
+ * @param matcher - matcher object containing fields and values to match against
+ * @returns
+ */
+export const buildMatcherScript = (matcher?: Matcher): estypes.Script => {
+  if (!matcher || matcher.fields.length === 0 || matcher.values.length === 0) {
+    throw new Error('Invalid matcher: fields and values must be defined and non-empty');
+  }
+
+  return {
+    lang: 'painless',
+    params: {
+      matcher_fields: matcher.fields,
+      matcher_values: matcher.values,
+    },
+    source: `
+      for (def f : params.matcher_fields) {
+        if (doc.containsKey(f) && !doc[f].empty) {
+          for (def v : doc[f]) {
+            if (params.matcher_values.contains(v)) return true;
+          }
+        }
+      }
+      return false;
+    `,
+  };
+};
+
+/**
+ * Building privileged search body for matchers
+ */
+export const buildPrivilegedSearchBody = (
+  script: estypes.Script,
+  timeGte: string = 'now-10y',
+  afterKey?: AfterKey,
+  pageSize: number = 100
+): Omit<estypes.SearchRequest, 'index'> => ({
+  size: 0,
+  query: {
+    range: { '@timestamp': { gte: timeGte } },
+  },
+  aggs: {
+    privileged_user_status_since_last_run: {
+      composite: {
+        size: pageSize,
+        sources: [{ username: { terms: { field: 'user.name' } } }],
+        ...(afterKey ? { after: afterKey } : {}),
+      },
+      aggs: {
+        latest_doc_for_user: {
+          top_hits: {
+            size: 1,
+            sort: [{ '@timestamp': { order: 'desc' as estypes.SortOrder } }],
+            script_fields: { 'user.is_privileged': { script } },
+            _source: { includes: ['user', 'roles', '@timestamp'] },
+          },
+        },
+      },
+    },
+  },
+});
+
+export const applyPrivilegedUpdates = async ({
+  dataClient,
+  users,
+}: {
+  dataClient: PrivilegeMonitoringDataClient;
+  users: PrivMonIntegrationsUser[];
+}) => {
+  if (users.length === 0) return;
+
+  const chunkSize = 500;
```

Comment:
I would suggest extracting all  `chunkSize`and `pageSize` configurations to a constant file, so we can easily update them.